x-pack/plugins/ingest_manager/server/services/epm/packages/install.ts

```diff
@@ -92,127 +92,113 @@ export async function installPackage(options: {
+
   const reinstall = pkgVersion === installedPkg?.attributes.version;
   const removable = !isRequiredPackage(pkgName);
   const { internal = false } = registryPackageInfo;
+  const toSaveESIndexPatterns = generateESIndexPatterns(registryPackageInfo.datasets);
 
-  // delete the previous version's installation's SO kibana assets before installing new ones
-  // in case some assets were removed in the new version
-  if (installedPkg) {
-    try {
-      await deleteKibanaSavedObjectsAssets(savedObjectsClient, installedPkg.attributes.installed);
-    } catch (err) {
-      // log these errors, some assets may not exist if deleted during a failed update
-    }
-  }
-
-  const [installedKibanaAssets, installedPipelines] = await Promise.all([
-    installKibanaAssets({
+  // add the package installation to the saved object
+  if (!installedPkg) {
+    await createInstallation({
       savedObjectsClient,
       pkgName,
       pkgVersion,
-      paths,
-    }),
-    installPipelines(registryPackageInfo, paths, callCluster),
-    // index patterns and ilm policies are not currently associated with a particular package
-    // so we do not save them in the package saved object state.
-    installIndexPatterns(savedObjectsClient, pkgName, pkgVersion),
-    // currenly only the base package has an ILM policy
-    // at some point ILM policies can be installed/modified
-    // per dataset and we should then save them
-    installILMPolicy(paths, callCluster),
-  ]);
+      internal,
+      removable,
+      installed_kibana: [],
+      installed_es: [],
+      toSaveESIndexPatterns,
+    });
+  }
 
-  // install or update the templates
+  const installIndexPatternPromise = installIndexPatterns(savedObjectsClient, pkgName, pkgVersion);
+  const installKibanaAssetsPromise = installKibanaAssets({
+    savedObjectsClient,
+    pkgName,
+    paths,
+    isUpdate,
+  });
+
+  // the rest of the installation must happen in sequential order
+
+  // currently only the base package has an ILM policy
+  // at some point ILM policies can be installed/modified
+  // per dataset and we should then save them
+  await installILMPolicy(paths, callCluster);
+
+  // installs versionized pipelines without removing currently installed ones
+  const installedPipelines = await installPipelines(
+    registryPackageInfo,
+    paths,
+    callCluster,
+    savedObjectsClient
+  );
+  // install or update the templates referencing the newly installed pipelines
   const installedTemplates = await installTemplates(
     registryPackageInfo,
+    isUpdate,
     callCluster,
-    pkgName,
-    pkgVersion,
-    paths
+    paths,
+    savedObjectsClient
   );
-  const toSaveESIndexPatterns = generateESIndexPatterns(registryPackageInfo.datasets);
 
+  // update current backing indices of each data stream
+  await updateCurrentWriteIndices(callCluster, installedTemplates);
+
+  // if this is an update, delete the previous version's pipelines
+  if (installedPkg && !reinstall) {
```

Comment:
do we need to use `isUpdate` here as well?

Reply:
Initially instead of `installedPkg` (which is the currently installed package if there is one) I had `isUpdate`, but I couldn't figure out how to get around typescript error `Object is possibly 'undefined'` in the `deletePipelines` function.  `isUpdate` checks that `installedPackage` exists at the start of the function so I don't know why it gives this error.  I should have written a note about it because this is confusing.  I will update the comments in a follow up PR until i figure out how to resolve the typescript issue.